### PR TITLE
[W4.4b][W13-A1]Raynold Ng

### PR DESCRIPTION
**Pull request by raynoldng**
LO-ImplementClasses

### CODE REVIEW

File: doc/UserGuide.md [hunk 1]
```diff
@@ -44,10 +44,10 @@ Format: `add NAME [p]p/PHONE_NUMBER [p]e/EMAIL [p]a/ADDRESS [t/TAG]...`
 > be seen using the `viewall` command.
 > 
 > Persons can have any number of tags (including 0)
-
+> Address must be in the following format: `BLOCK, STREET, UNIT, POSTALCODE`
```
Comment (okkhoy) [inline doc/UserGuide.md:47]: :+1:  for mentioning the format and giving appropriate examples in the user guide.

File: src/seedu/addressbook/data/person/Address.java [hunk 2]
```diff
@@ -1,18 +1,32 @@
...
 /**
  * Represents a Person's address in the address book.
  * Guarantees: immutable; is valid as declared in {@link #isValidAddress(String)}
  */
 public class Address {
 
-    public static final String EXAMPLE = "123, some street";
-    public static final String MESSAGE_ADDRESS_CONSTRAINTS = "Person addresses can be in any format";
-    public static final String ADDRESS_VALIDATION_REGEX = ".+";
+    // format: BLOCK, STREET, UNIT, POSTAL_CODE
```
Comment (okkhoy) [inline src/seedu/addressbook/data/person/Address.java:13]: remove unnecessary code completely; you can always recover it from previous versions RCS \#FTW

File: src/seedu/addressbook/data/person/Address.java [hunk 3]
```diff
@@ -1,18 +1,32 @@
...
+    public static final String EXAMPLE = "123, Clementi Ave 3, #12-34, 231534";
+    public static final String MESSAGE_ADDRESS_CONSTRAINTS = "Person addresses should be in the format of: " +
+            "'a/BLOCK, STREET, UNIT, POSTAL_CODE'";
+    public static final String ADDRESS_VALIDATION_REGEX = "(.+), (.+), (.+), (.+)";
 
+    private static final int BLOCK_INDEX = 1;
```
Comment (okkhoy) [inline src/seedu/addressbook/data/person/Address.java:19]: 2 things here:

- Good to see you have eliminated magic numbers in the code
- However, please take a look at the coding standards and find out what is the correct convention for naming constants

File: src/seedu/addressbook/data/person/Address.java [hunk 4]
```diff
@@ -1,18 +1,32 @@
...
+    private static final int STREET_INDEX = 2;
+    private static final int UNIT_INDEX = 3;
+    private static final int POSTAL_CODE_INDEX = 4;
+
+    private final Block block;
+    private final Street street;
+    private final Unit unit;
+    private final PostalCode postalCode;
     public final String value;
+
```
Comment (okkhoy) [inline src/seedu/addressbook/data/person/Address.java:29]: I notice that you are still retaining the `value`. Imagine the havoc duplicated data can create!

File: src/seedu/addressbook/data/person/AddressProperty.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
+package seedu.addressbook.data.person;
+
+/**
+ * Created by raynoldng on 5/2/18.
```
Comment (okkhoy) [inline src/seedu/addressbook/data/person/AddressProperty.java:4]: We can get to know this from the commit history; as such, remove these kinda redundant/not-so-useful comments generated by the IDE.

File: src/seedu/addressbook/data/person/AddressProperty.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ */
+public class AddressProperty {
```
Comment (okkhoy) [inline src/seedu/addressbook/data/person/AddressProperty.java:6]: This class and its non-trivial methods should have header comments. 
If not, developers who write client code for this class will not know how to use it.

File: src/seedu/addressbook/data/person/Block.java [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package seedu.addressbook.data.person;
+
+/**
+ * Created by raynoldng on 5/2/18.
+ */
+public class Block extends AddressProperty {
+    public Block (String value) {
+        super(value);
```
Comment (okkhoy) [inline src/seedu/addressbook/data/person/Block.java:8]: You can consider moving the validation and equality check to the new classes you have created.

File: src/seedu/addressbook/data/person/AddressProperty.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
+package seedu.addressbook.data.person;
```
Comment (okkhoy) [inline src/seedu/addressbook/data/person/AddressProperty.java:1]: BTW, this is a unique attempt I have seen in so many years :+1:

File: test/expected.txt [hunk 9]
```diff
@@ -85,57 +85,57 @@
 || Enter command: || [Command entered:  add Valid Name p/12345 e/valid@e.mail a/valid, address t/invalid_-[.tag]
 || Tags names should be alphanumeric
 || ===================================================
-|| Enter command: || [Command entered:  add Adam Brown p/111111 e/adam@gmail.com a/111, alpha street]
-|| New person added: Adam Brown Phone: 111111 Email: adam@gmail.com Address: 111, alpha street Tags: 
+|| Enter command: || [Command entered:  add Adam Brown p/111111 e/adam@gmail.com a/111, alpha street, #12-34, 231534]
+|| New person added: Adam Brown Phone: 111111 Email: adam@gmail.com Address: 111, alpha street, #12-34, 231534 Tags:
```
Comment (okkhoy) [inline test/expected.txt:89]: it is good to see you have updated the tests as well.

**Review by okkhoy — COMMENTED** [review]
This is generally a good attempt!

**Comment by okkhoy** [comment]
@raynoldng 
Some comments added. Please ack & close the PR after reading comments.

**Comment by raynoldng** [comment]
Noted with thanks!

**Pull request closed without merging**